sdk-utils/src/main/java/com/redhat/parodos/sdkutils/RetryExecutorService.java

```diff
@@ -37,48 +39,41 @@ public T submitWithRetry(Callable<T> task) {
 	 */
 	public T submitWithRetry(Callable<T> task, Runnable onSuccess, Runnable onFailure, long maxRetryTime,
 			long retryDelay) {
-		Future<T> future = executor.submit(() -> {
-			long startTime = System.currentTimeMillis();
-			long endTime = startTime + maxRetryTime;
+		CompletableFuture<T> future = new CompletableFuture<>();
+		long startTime = System.currentTimeMillis();
+		long endTime = startTime + maxRetryTime;
 
-			while (System.currentTimeMillis() < endTime) {
-				try {
-					T result = task.call();
-					onSuccess.run();
-					return result; // Success, no need to retry
-				}
-				catch (Exception e) {
-					// Task failed, invoke onFailure callback
-					onFailure.run();
-
-					// Sleep for the retry delay
-					try {
-						// FIXME: This is a blocking call, we should use a non-blocking
-						// sleep
-						Thread.sleep(retryDelay);
-					}
-					catch (InterruptedException ex) {
-						Thread.currentThread().interrupt();
-						return null; // Interrupted, exit the task
-					}
-				}
+		ScheduledFuture<?> scheduledFuture = scheduledExecutor.scheduleWithFixedDelay(() -> {
+			if (System.currentTimeMillis() >= endTime) {
+				future.completeExceptionally(new TimeoutException("Retry limit reached."));
+				return;
 			}
 
-			return null; // Retry limit reached
-		});
+			try {
+				T result = task.call();
+				onSuccess.run();
+				future.complete(result); // Success, complete the future with the result
+			}
+			catch (Exception e) {
+				onFailure.run();
+			}
+		}, 0, retryDelay, TimeUnit.MILLISECONDS);
 
 		try {
 			return future.get();
 		}
 		catch (InterruptedException | ExecutionException e) {
-			throw new RuntimeException(e);
+			return null;
```

Comment:
Can you explain me, why we return `null` here?

Reply:
We should add unit-test to our unit-test :-)